pkg/code-generator/codegen/templates/snapshot_emitter_template.go

```diff
@@ -171,74 +202,279 @@ func (c *{{ lower_camel .GoName }}Emitter) Snapshots(watchNamespaces []string, o
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "starting {{ upper_camel .Name }} watch")
+			}
 
+			done.Add(1)
+			go func(namespace string) {
+				defer done.Done()
+				errutils.AggregateErrs(ctx, errs, {{ lower_camel .Name }}Errs, namespace+"-{{ lower_camel .PluralName }}")
+			}(namespace)
 {{- end }}
-{{- end}}
+{{- end }}
+			/* Watch for changes and update snapshot */
+			go func(namespace string) {
+				for {
+					select {
+					case <-ctx.Done():
+						return
+{{- range .Resources }}
+{{- if (not .ClusterScoped) }}
+					case {{ lower_camel .Name }}List, ok := <-{{ lower_camel .Name }}NamespacesChan:
+						if !ok {
+							return
+						}
+						select {
+						case <-ctx.Done():
+							return
+						case {{ lower_camel .Name }}Chan <- {{ lower_camel .Name }}ListWithNamespace{list: {{ lower_camel .Name }}List, namespace: namespace}:
+						}
+{{- end }}
+{{- end }}
+					}
+				}
+			}(namespace)
+		}
+		// create watch on all namespaces, so that we can add resources from new namespaces
+		// TODO-JAKE this interface has to deal with the event types of kubernetes independently without the interface knowing about it.
+		// we will need a way to deal with DELETES and CREATES and updates seperately
+		namespaceWatch, _, err := c.resourceNamespaceLister.GetNamespaceResourceWatch(ctx, resources.ResourceNamespaceWatchOptions{
+			FieldSelectors: excludeNamespacesFieldDesciptors,
+			ExpressionSelectors: opts.ExpressionSelector,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
 
-		/* Watch for changes and update snapshot */
-		go func(namespace string) {
+		go func() {
 			for {
 				select {
 				case <-ctx.Done():
 					return
-{{- range .Resources}}
-{{- if (not .ClusterScoped) }}
-				case {{ lower_camel .Name }}List, ok := <- {{ lower_camel .Name }}NamespacesChan:
+				case resourceNamespaces, ok := <-namespaceWatch:
 					if !ok {
 						return
 					}
-					select {
-					case <-ctx.Done():
-						return
-					case {{ lower_camel .Name }}Chan <- {{ lower_camel .Name }}ListWithNamespace{list:{{ lower_camel .Name }}List, namespace:namespace}:
+					// TODO-JAKE with the interface, we have lost the ability to know the event type.
+					// so the interface must be able to identify the type of event that occured as well
+					// not just return the list of namespaces
+					newNamespaces := []string{}
+
+					for _, ns := range resourceNamespaces {
+						// TODO-JAKE are we sure we need this. Looks like there is a cocurrent map read and map write here
+						
+{{- range .Resources }}
+{{- if (not .ClusterScoped) }}
+						// TODO-JAKE we willl only need to do this once, I might be best to keep a set/map of the current
+						// namespaces that are used
+						if _, hit := {{ lower_camel .PluralName }}ByNamespace.Load(ns.Name); !hit {
+							newNamespaces = append(newNamespaces, ns.Name)
```

Comment:
This could be a problem if we have a lot of resources, I am thinking of adding a map of visited namespaces, or at least namespaces with watches currently on them.

Reply:
Where are the 2 logs? I am not sure what you mean by this? we can have a conversation on this.

Reply:
The 2 logs were just where we hit it in here plus the aggregator at the calling level.